# Telco Customer Churn
### Introduction

Created by: Wan Nazirul, Feb 2017

As part of fullfilment in UTM BIA Data Mining course, this assingment focus on basic data exploration using pyhton. 

### Step1. Import neccessary library

In [2]:
import pandas as pd
import numpy as np

### Step 2. Import the dataset from this [address](https://drive.google.com/file/d/0ByCqHmF4VJXiTkRzeE1yVjZvVEk/view) and assign it to data frame.

*There is no metadata available explaining the meaning and function of each column

In [8]:
data = pd.read_csv('Telco-Customer-Churn.csv')

### Step 3. Summarising the DataFrame

Numbers of row, column and data type

In [13]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
customerID          7043 non-null object
gender              7043 non-null object
SeniorCitizen       7043 non-null int64
Partner             7043 non-null object
Dependents          7043 non-null object
tenure              7043 non-null int64
PhoneService        7043 non-null object
MultipleLines       7043 non-null object
InternetService     7043 non-null object
OnlineSecurity      7043 non-null object
OnlineBackup        7043 non-null object
DeviceProtection    7043 non-null object
TechSupport         7043 non-null object
StreamingTV         7043 non-null object
StreamingMovies     7043 non-null object
Contract            7043 non-null object
PaperlessBilling    7043 non-null object
PaymentMethod       7043 non-null object
MonthlyCharges      7043 non-null float64
TotalCharges        7043 non-null object
Churn               7043 non-null object
dtypes: float64(1), int64(2), obj

View the first and last 5 row

In [14]:
data.head(5)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ...  DeviceProtection  \
0  No phone service             DSL             No  ...                No   
1                No             DSL            Yes  ...               Yes   
2                No             DSL            Yes  ...                No   
3  No phone service             DSL            Yes  ...               Yes   
4                No     Fiber optic             No  ...                No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [15]:
data.tail(5)

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
7038  6840-RESVB    Male              0     Yes        Yes      24   
7039  2234-XADUH  Female              0     Yes        Yes      72   
7040  4801-JZAZL  Female              0     Yes        Yes      11   
7041  8361-LTMKD    Male              1     Yes         No       4   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...   \
7038          Yes               Yes             DSL            Yes  ...    
7039          Yes               Yes     Fiber optic             No  ...    
7040           No  No phone service             DSL            Yes  ...    
7041          Yes               Yes     Fiber optic             No  ...    
7042          Yes                No     Fiber optic            Yes  ...    

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

     PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
7038              Yes               Mailed check          84.80        1990.5   
7039              Yes    Credit card (automatic)         103.20        7362.9   
7040              Yes           Electronic check          29.60        346.45   
7041              Yes               Mailed check          74.40         306.6   
7042              Yes  Bank transfer (automatic)         105.65        6844.5   

     Churn  
7038    No  
7039    No  
7040    No  
7041   Yes  
7042    No  

[5 rows x 21 columns]

### Step 4. Exploring data

* Count and percentage of churn by Internet Service type

In [63]:
churn_intServ = data.groupby(['InternetService','Churn'])['Churn'].count()
churn_intServ

InternetService  Churn
DSL              No       1962
                 Yes       459
Fiber optic      No       1799
                 Yes      1297
No               No       1413
                 Yes       113
Name: Churn, dtype: int64

In [64]:
np.round(churn_intServ.groupby(level=0).apply(lambda x:100 * x / float(x.sum())) ,2)

InternetService  Churn
DSL              No       81.04
                 Yes      18.96
Fiber optic      No       58.11
                 Yes      41.89
No               No       92.60
                 Yes       7.40
Name: Churn, dtype: float64

Almost half of customers using ***Fiber Optic*** have churn tendency as high as **41.89%** while customer using ***DSL*** are most loyalty with percentage of staying with the same telco as high as **81.04%**.

In [72]:
np.round(data.groupby(['InternetService','Churn'])['tenure'].mean(),2)

InternetService  Churn
DSL              No       37.20
                 Yes      14.11
Fiber optic      No       42.09
                 Yes      20.20
No               No       32.33
                 Yes       8.24
Name: tenure, dtype: float64

In [73]:
data.groupby(['InternetService','Churn'])['tenure'].mean()

InternetService  Churn
DSL              No       37.198777
                 Yes      14.111111
Fiber optic      No       42.089494
                 Yes      20.196608
No               No       32.331210
                 Yes       8.238938
Name: tenure, dtype: float64

In [78]:
test = data['tenure'] > 30
test.count()

7043

In [83]:
#f1.loc[:, df1.loc['a'] > 0]
data[data['tenure'] > 20.2].count()

customerID          4165
gender              4165
SeniorCitizen       4165
Partner             4165
Dependents          4165
tenure              4165
PhoneService        4165
MultipleLines       4165
InternetService     4165
OnlineSecurity      4165
OnlineBackup        4165
DeviceProtection    4165
TechSupport         4165
StreamingTV         4165
StreamingMovies     4165
Contract            4165
PaperlessBilling    4165
PaymentMethod       4165
MonthlyCharges      4165
TotalCharges        4165
Churn               4165
dtype: int64